File: Cache/Clip.swift
import Foundation
import SwiftData

/// One captured pasteboard item.
///
/// `@Model` is SwiftData's macro. It rewrites this class at compile time so every
/// stored property is backed by the database — you write plain Swift, SwiftData
/// handles persistence. No schema file, no migration boilerplate for now.
///
/// Phase A only populates the text fields. The image/OCR/category fields are
/// declared now so Phase B and D don't require a schema migration.
@Model
final class Clip {
    var id: UUID
    var createdAt: Date

    /// "text" | "link" | "image" | "color" | "code" | "file".
    /// Stored as String rather than an enum because SwiftData handles primitives
    /// most predictably, and it keeps the schema stable if we add kinds later.
    var kind: String

    /// Text content, or a file path string, or a hex code — depending on `kind`.
    var text: String?

    var imageFilename: String?
    var thumbnailFilename: String?
    var ocrText: String?

    var sourceAppName: String?
    var sourceAppBundleID: String?

    var isPinned: Bool

    /// User-created category name. nil means it lives in plain History.
    var category: String?

    // MARK: - Added after the first schema
    //
    // Every property below is optional or carries a default. That's what lets
    // SwiftData migrate an existing store in place instead of demanding a
    // versioned migration plan — and it's why an existing history survives an
    // app update rather than being wiped.

    /// User-supplied name for the clip, shown instead of its contents.
    var title: String?

    /// How many times this clip has been pasted back. Drives "most used" sort.
    var useCount: Int = 0

    /// Last time it was pasted back, for recency sorting.
    var lastUsedAt: Date?

    /// When set, a local notification fires at this time.
    var reminderAt: Date?

    /// Position for manual sort order. Unset until the user drags something.
    var manualOrder: Int?

    init(
        id: UUID = UUID(),
        createdAt: Date = Date(),
        kind: String,
        text: String? = nil,
        imageFilename: String? = nil,
        thumbnailFilename: String? = nil,
        ocrText: String? = nil,
        sourceAppName: String? = nil,
        sourceAppBundleID: String? = nil,
        isPinned: Bool = false,
        category: String? = nil,
        title: String? = nil,
        useCount: Int = 0,
        lastUsedAt: Date? = nil,
        reminderAt: Date? = nil,
        manualOrder: Int? = nil
    ) {
        self.id = id
        self.createdAt = createdAt
        self.kind = kind
        self.text = text
        self.imageFilename = imageFilename
        self.thumbnailFilename = thumbnailFilename
        self.ocrText = ocrText
        self.sourceAppName = sourceAppName
        self.sourceAppBundleID = sourceAppBundleID
        self.isPinned = isPinned
        self.category = category
        self.title = title
        self.useCount = useCount
        self.lastUsedAt = lastUsedAt
        self.reminderAt = reminderAt
        self.manualOrder = manualOrder
    }
}

File: windows/src/Cache.Core/Clip.cs
namespace Cache.Core;

/// <summary>
/// One captured clipboard item.
///
/// Field-for-field the same shape as the macOS <c>Clip</c> SwiftData model, on
/// purpose: when sync eventually happens, the two stores should describe the
/// same thing rather than needing a mapping layer invented after the fact.
/// </summary>
public sealed class Clip
{
    public required Guid Id { get; init; }

    /// <summary>UTC. Stored as an ISO-8601 string so the DB stays human-readable.</summary>
    public required DateTimeOffset CreatedAt { get; init; }

    public required ClipKind Kind { get; set; }

    /// <summary>Text content, a newline-joined file path list, or a hex code — depending on <see cref="Kind"/>.</summary>
    public string? Text { get; set; }

    public string? ImageFilename { get; set; }
    public string? ThumbnailFilename { get; set; }

    /// <summary>
    /// Text recognised inside an image. Null means recognition never ran; empty
    /// string means it ran and found nothing. Backfill relies on that
    /// distinction to avoid re-scanning textless images on every launch.
    /// </summary>
    public string? OcrText { get; set; }

    public string? SourceAppName { get; set; }

    /// <summary>
    /// On Windows this is the executable path of the foreground process, which
    /// is the closest equivalent to a macOS bundle identifier.
    /// </summary>
    public string? SourceAppId { get; set; }

    public bool IsPinned { get; set; }

    /// <summary>User-created collection name. Null means plain History.</summary>
    public string? Category { get; set; }

    /// <summary>User-supplied name, shown instead of the contents.</summary>
    public string? Title { get; set; }

    /// <summary>How many times this clip has been pasted back. Drives "most used" sort.</summary>
    public int UseCount { get; set; }

    public DateTimeOffset? LastUsedAt { get; set; }
    public DateTimeOffset? ReminderAt { get; set; }

    /// <summary>Position for manual sort order. Null until the user drags something.</summary>
    public int? ManualOrder { get; set; }

    /// <summary>What the list shows for this clip when it has no title of its own.</summary>
    public string Preview => Title
        ?? Kind switch
        {
            ClipKind.Image => string.IsNullOrWhiteSpace(OcrText) ? "Image" : OcrText.Trim(),
            _ => Text?.Trim() ?? string.Empty,
        };
}
